proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,188 @@
+apply to *new* modules.
+
+* Over time, and only with the agreement and support of the Core Libraries Committee, we may remove some ``GHC.*`` modules
+  from ``base``, especially ones that are barely used, or are manifestly "internal" (i.e. part of the implementation
+  of other, more public functions).
+  Of course there would be a significant deprecation cycle, to allow client libraries to adapt.
+
+Proposal 3 only expresses a direction of travel.  We will have to see what the CLC's attitude is,
+and what the Haskell community thinks.  Anything that disturbs the API of base needs to be considered
+rather carefully.
+
+
+Proposal 4
+------------
+
+All of the modules in package ``ghc`` currently start with ``GHC.*`` which
+(currently correctly) signals that they are part of GHC's internals.
+
+As part of the GHC API redesign (a HF project in its own right, currently stalled) it would be very helpful
+to identify a (multi-module) stable API for package ``ghc``. In that way, users of package ``ghc``
+could know whether
+they are using a curated, relatively-stable API function, or reaching deep into GHC's guts and using
+a random fuction whose name or type, or very existence, might change without warning. Hence:
+
+* The public API of package ``ghc`` (GHC as a library) should have modules whose names clearly distinguish them
+  from internal modules.
+
+For example, the public API could have modules of form ``GhcAPI.*``, or ``GHC.API.*``, or ``Language.Haskell.GHC.*`` or something else. The specifics are a matter for the future GHC API working group.
+
+
+
+Timescale
+==========
+The first release of GHC with ``ghc-experimental`` and ``ghc-internal`` will be GHC 9.10, which expect to
+release in early 2024.  It would be good to establish naming conventions for modules well before this date.
+
+Example lifecycle
+===================
+
+By way of example, consider the ``HasField`` class, which supports overloaded record fields.
+It is currently defined in ``base:GHC.Records``, which is an odd module to have to import.
+Moreover there is
+more than one GHC proposal that suggest changes to its design (e.g. see `GHC Proposal 158 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0158-record-set-field.rst>`_); it is not nearly as stable as most of ``base``
+
+If ``ghc-experimental`` had existed we would have put it in ``ghc-experimental:Data.Records.Experimental``.
+That would have made it clear that the design of overloaded records still evolving.
+Once the design becomes settled and stable, it could move to ``base``, perhaps in a module like ``Data.Records``.
+
+Other similar examples include
+
+* The tuple proposal of `GHC Proposal 475 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0475-tuple-syntax.rst>`_
+* The `DataToTag CLC proposal <https://github.com/haskell/core-libraries-committee/issues/104>`_ would have been easier to expose through ``ghc-experimental`` in the first instance.
```

Comment:
I don't think this is a good example; the motivation for that proposal is not convincing unless the _existing_ `dataToTag#` and `getTag` exposed from `ghc-prim` and `base` get improved types.

Reply:
OK. I have replaced this example with the exceptions proposal.